syftlib/src/main/java/org/openmined/kotlinsyft/WebRTC.kt

```diff
@@ -0,0 +1,304 @@
+                        worker_id
+                    )
+                    sendIce()
+                } else {
+                    // We've just set remote SDP - do nothing for now -
+                    // answer will be created soon.
+                    Log.d(TAG, "successfully finished setting offer as Remote description")
+                }
+
+            }
+        }
+
+        override fun onCreateSuccess(sessionDescription: SessionDescription) {
+            // This is called when answer or offer is created
+            Log.d(TAG, "created")
+            if (peers[worker_id]?.connection != null && peers[worker_id]?.connection?.localDescription != null) {
+                Log.e(TAG, "multiple SDP creation")
+                return
+            }
+            peers[worker_id]?.connection?.setLocalDescription(this, sessionDescription)
+        }
+
+        override fun onCreateFailure(error: String?) {
+            Log.e(TAG, "error creating : $error")
+        }
+
+    }
+
+    inner class PeerConnectionObserver(
+        private val worker_id: String,
+        private val creatorType: SDP_type
+    ) : Observer {
+
+        override fun onIceCandidate(new_candidate: IceCandidate?) {
+            if (new_candidate != null) {
+                Log.d(TAG, "Saving new ICE Candidate")
+                peers[worker_id]?.candidateQueue?.add(new_candidate)
+            }
+        }
+
+        override fun onDataChannel(dc: DataChannel) {
+            Log.d(TAG, "Calling onDataChannel ${dc.label()}")
+            dc.registerObserver(DataChannelObserver(dc))
+            peers[worker_id]?.channel = dc
+
+        }
+
+        override fun onIceConnectionReceivingChange(p0: Boolean) {
+            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        }
+
+        override fun onIceConnectionChange(p0: IceConnectionState?) {
+            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        }
+
+        override fun onIceGatheringChange(gatheringState: IceGatheringState) {}
+
+        override fun onSignalingChange(signalingState: SignalingState) {
+            Log.d(TAG, "Signalling State: $signalingState")
+        }
+
+        override fun onIceCandidatesRemoved(p0: Array<out IceCandidate>?) {
+            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        }
+
+        override fun onRenegotiationNeeded() {
+            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+        }
+
+        override fun onAddTrack(p0: RtpReceiver?, p1: Array<out MediaStream>?) {}
+        override fun onRemoveStream(p0: MediaStream?) {}
+        override fun onAddStream(p0: MediaStream?) {}
+
+    }
+
+    class DataChannelObserver(private val dataChannel: DataChannel?) : DataChannel.Observer {
```

Comment:
Maybe extract this class and the ones below to separate files. It looks as if this file could grow too much.

Reply:
I initially thought of shifting the observers to a different file but then we would need to send the reference of WebRTC object. The issue is the observer itself is created by WebRTC and sending references to itself via peer references forms a syntactic loop. I wanted to avoid that and make reference implicit

Reply:
If it starts to grow we can then shift them out

Reply:
I was considering this but then referencing WebRTC instance into observer which itself is created by webRTC internally maintaining a relationship with the connection. An observer, in that case, should rather have an implicit reference to its creator otherwise it just forms a syntactic loop

Only message chunking would get added later I believe whose code will not be added her